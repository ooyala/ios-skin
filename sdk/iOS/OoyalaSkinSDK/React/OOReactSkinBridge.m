//
//  OOReactSkinBridge.m
//  OoyalaSkinSDK
//
//  Created on 8/14/18.
//  Copyright Â© 2018 ooyala. All rights reserved.
//

#import "OOReactSkinBridge.h"
#import "OOReactSkinBridgeModule.h"
#import <React/RCTBridge+Private.h>

@interface OOReactSkinBridge ()

@property (nonatomic, weak, readonly) id<OOReactSkinBridgeDelegate> delegate;

@end


@implementation OOReactSkinBridge

@dynamic delegate;

- (instancetype)initWithDelegate:(id<OOReactSkinBridgeDelegate>)delegate
                   launchOptions:(NSDictionary *)launchOptions {
  return [super initWithDelegate:delegate launchOptions:launchOptions];
}

- (void)invalidate {
  [super invalidate];
  [NSNotificationCenter.defaultCenter removeObserver:self];
}

- (void)setUp {
  [super setUp];
  [NSNotificationCenter.defaultCenter addObserver:self
                                         selector:@selector(didInitializeModuleNotification:)
                                             name:RCTDidInitializeModuleNotification
                                           object:self.batchedBridge];
}

- (void)didInitializeModuleNotification:(NSNotification *)notification {
  id module = notification.userInfo[@"module"];

  if ([self.delegate respondsToSelector:@selector(bridge:didLoadModule:)]) {
    if (module && [module conformsToProtocol:@protocol(OOReactSkinBridgeModule)]) {
      [self.delegate bridge:self didLoadModule:module];
    }
  }
}

@end


@implementation RCTBridge (OOReactSkinEventsEmitterable)

- (OOReactSkinEventsEmitter *)skinEventsEmitter {
  return [self moduleForName:[OOReactSkinEventsEmitter moduleName]];
}

@end
