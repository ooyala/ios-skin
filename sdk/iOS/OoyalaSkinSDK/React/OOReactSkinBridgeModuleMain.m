//
//  OOReactSkinBridgeModuleMain.m
//  OoyalaSkinSDK
//
//  Created by Maksim Kupetskii on 8/14/18.
//  Copyright Â© 2018 ooyala. All rights reserved.
//

#import "OOReactSkinBridgeModuleMain.h"
#import "NSMutableDictionary+Utils.h"

#import <OoyalaSDK/OODebugMode.h>
#import <OoyalaSDK/OOOoyalaPlayer.h>

@interface OOReactSkinBridgeModuleMain ()

@property (nonatomic, weak, nullable) id<OOReactSkinModelDelegate> skinModelDelegate;

@end


@implementation OOReactSkinBridgeModuleMain

@synthesize bridge;

RCT_EXPORT_MODULE(OOReactSkinBridgeModuleMain);

- (dispatch_queue_t)methodQueue {
  return dispatch_get_main_queue();
}

- (OOReactSkinBridgeModuleType)moduleType {
  return OOReactSkinBridgeModuleTypeMain;
}

- (void)setSkinViewDelegate:(nonnull id<OOReactSkinModelDelegate>)delegate {
  _skinModelDelegate = delegate;
}

#pragma mark - Constants
#pragma mark Buttons
static NSString *playPauseButtonName    = @"PlayPause";
static NSString *playButtonName         = @"Play";
static NSString *rewindButtonName       = @"rewind";
static NSString *socialShareButtonName  = @"SocialShare";
static NSString *fullscreenButtonName   = @"fullscreen";
static NSString *learnMoreButtonName    = @"LearnMore";
static NSString *skipButtonName         = @"Skip";
static NSString *moreOptionButtonName   = @"More";
static NSString *upNextDismiss          = @"upNextDismiss";
static NSString *upNextClick            = @"upNextClick";
static NSString *adIconButtonName       = @"Icon";
static NSString *adOverlayButtonName    = @"Overlay";
static NSString *stereoscopicButtonName = @"stereoscopic";
static NSString *pipButtonName          = @"PIP";
static NSString *replayButtonName       = @"replay";

#pragma mark Keys
static NSString *nameKey              = @"name";
static NSString *embedCodeKey         = @"embedCode";
static NSString *percentageKey        = @"percentage";
static NSString *languageKey          = @"language";
static NSString *bucketInfoKey        = @"bucketInfo";
static NSString *actionKey            = @"action";
static NSString *audioTrackKey        = @"audioTrack";
static NSString *indexKey             = @"index";
static NSString *clickUrlKey          = @"clickUrl";
static NSString *eventNameKey         = @"eventName";
static NSString *playbackSpeedRateKey = @"playbackSpeedRate";
static NSString *volumeKey            = @"volume";

#pragma mark Values
static NSString *startValue   = @"start";
static NSString *moveValue    = @"move";
static NSString *endValue     = @"end";
static NSString *clickValue   = @"click";
static NSString *impressValue = @"impress";

#pragma mark - Exported Methods

RCT_EXPORT_METHOD(onVolumeChanged:(NSDictionary *)parameters) {
  float volume = [parameters[volumeKey] floatValue];
  [self.skinModelDelegate handleVolumeChanged:volume];
}

RCT_EXPORT_METHOD(onMounted) {
  LOG(@"onMounted - Not going to use at the moment");
}

RCT_EXPORT_METHOD(onLanguageSelected:(NSDictionary *)parameters) {
  [self.skinModelDelegate handleLanguageSelection:parameters[languageKey]];
}

RCT_EXPORT_METHOD(onAudioTrackSelected:(NSDictionary *)parameters) {
  [self.skinModelDelegate handleAudioTrackSelection:parameters[audioTrackKey]];
}

RCT_EXPORT_METHOD(onPlaybackSpeedRateSelected:(NSDictionary *)parameters) {
  [self.skinModelDelegate handlePlaybackSpeedRateSelection:parameters[playbackSpeedRateKey]];
}

RCT_EXPORT_METHOD(onPress:(NSDictionary *)parameters) {
  NSString *buttonName = parameters[nameKey];

  if ([buttonName isEqualToString:playPauseButtonName]) {
    [self.skinModelDelegate handlePlayPause];
  } else if([buttonName isEqualToString:playButtonName]) {
    [self.skinModelDelegate handlePlay];
  } else if([buttonName isEqualToString:replayButtonName]) {
    [self.skinModelDelegate handleReplay];
  } else if([buttonName isEqualToString:rewindButtonName]) {
    [self.skinModelDelegate handleRewind];
  } else if([buttonName isEqualToString:socialShareButtonName]) {
    [self.skinModelDelegate handleSocialShare];
  } else if([buttonName isEqualToString:fullscreenButtonName]) {
    [self.skinModelDelegate toggleFullscreen];
  } else if([buttonName isEqualToString:learnMoreButtonName]) {
    [self.skinModelDelegate handleLearnMore];
  } else if([buttonName isEqualToString:skipButtonName]) {
    [self.skinModelDelegate handleSkip];
  } else if ([buttonName isEqualToString:adIconButtonName]) {
    [self.skinModelDelegate handleIconClick:[parameters[indexKey] integerValue]];
  } else if([buttonName isEqualToString:moreOptionButtonName]) {
    [self.skinModelDelegate handleMoreOption];
  } else if([buttonName isEqualToString:upNextDismiss]) {
    [self.skinModelDelegate handleUpNextDismiss];
  } else if([buttonName isEqualToString:upNextClick]) {
    [self.skinModelDelegate handleUpNextClick];
  } else if ([buttonName isEqualToString:adOverlayButtonName]) {
    [self.skinModelDelegate handleOverlay:parameters[clickUrlKey]];
  } else if ([buttonName isEqualToString:pipButtonName]) {
    [self.skinModelDelegate handlePip];
  } else if ([buttonName isEqualToString:stereoscopicButtonName]) {
    [self.skinModelDelegate toggleStereoMode];
  }
}

RCT_EXPORT_METHOD(handleTouchStart:(NSDictionary *)params){
  NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithDictionary:params];
  [result mergeWith:@{eventNameKey: startValue}];
  [self.skinModelDelegate handleTouch:result];
}

RCT_EXPORT_METHOD(handleTouchMove:(NSDictionary *)params) {
  NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithDictionary:params];
  [result mergeWith:@{eventNameKey: moveValue}];
  [self.skinModelDelegate handleTouch:result];
}

RCT_EXPORT_METHOD(handleTouchEnd:(NSDictionary *)params) {
  NSMutableDictionary *result = [[NSMutableDictionary alloc] initWithDictionary:params];
  [result mergeWith:@{eventNameKey: endValue}];
  [self.skinModelDelegate handleTouch:result];
}

RCT_EXPORT_METHOD(onScrub:(NSDictionary *)parameters) {
  Float64 position = [parameters[percentageKey] doubleValue];
  [self.skinModelDelegate handleScrub:position];
}

RCT_EXPORT_METHOD(setEmbedCode:(NSDictionary *)parameters) {
  NSString *embedCode = parameters[embedCodeKey];
  [self.skinModelDelegate setEmbedCode:embedCode];
}

RCT_EXPORT_METHOD(onDiscoveryRow:(NSDictionary *)parameters) {
  NSString *action     = parameters[actionKey];
  NSString *bucketInfo = parameters[bucketInfoKey];

  if ([action isEqualToString:clickValue]) {
    NSString *embedCode = parameters[embedCodeKey];
    [self.skinModelDelegate handleDiscoveryClick:bucketInfo embedCode:embedCode];
  } else if ([action isEqualToString:impressValue]) {
    [self.skinModelDelegate handleDiscoveryImpress:bucketInfo];
  }
}

@end
