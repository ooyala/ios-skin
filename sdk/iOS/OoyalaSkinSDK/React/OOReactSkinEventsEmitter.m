/**
 * @file       OOReactSkinEventsEmitter.m
 * @class      OOReactSkinEventsEmitter OOReactSkinEventsEmitter.m "OOReactSkinEventsEmitter.m"
 * @brief      OOReactSkinEventsEmitter
 * @details    OOReactSkinEventsEmitter.m in OoyalaSDK
 * @date       8/13/18
 * @copyright Copyright (c) 2018 Ooyala, Inc. All rights reserved.
 */

#import "OOReactSkinEventsEmitter.h"
#import "OOConstant.h"
#import "OOQueuedEvent.h"

#import <OoyalaSDK/OOOoyalaPlayer.h>
#import <OoyalaSDK/OOClosedCaptions.h>
#import <OoyalaSDK/OODebugMode.h>

@interface OOReactSkinEventsEmitter ()

@property (atomic) NSMutableArray *queuedEvents; //QueuedEvent *

@end

@implementation OOReactSkinEventsEmitter

RCT_EXPORT_MODULE(OOReactSkinEventsEmitter);

+ (BOOL)requiresMainQueueSetup {
  return YES;
}

- (instancetype)init {
  if (self = [super init]) {
    _isReactReady = NO;
    _queuedEvents = [NSMutableArray new];
  }
  return self;
}

- (NSArray<NSString *> *)supportedEvents {
  return @[@"timeChanged",
           @"seekStarted",
           @"seekCompleted",
           @"ccStylingChanged",
           @"currentItemChanged",
           @"frameChanged",
           @"fullscreenToggled",
           @"volumeChanged",
           @"playCompleted",
           @"stateChanged",
           @"desiredStateChanged",
           @"discoveryResultsReceived",
           @"onClosedCaptionUpdate",
           @"adStarted",
           @"adSwitched",
           @"adPodStarted",
           @"adPodCompleted",
           @"adOverlay",
           @"adError",
           @"setNextVideo",
           @"upNextDismissed",
           @"playStarted",
           @"error",
           @"embedCodeSet",
           @"controllerKeyPressEvent",
           @"vrContentEvent",
           @"multiAudioEnabled",
           @"audioTrackChanged",
           @"playbackSpeedEnabled",
           @"playbackSpeedRateChanged"];
}

- (void)sendDeviceEventWithName:(NSString *)eventName body:(id)body {
  if (self.isReactReady) {
    if (![eventName isEqualToString:OOOoyalaPlayerTimeChangedNotification] &&
        ![eventName isEqualToString:OO_CLOSED_CAPTIONS_UPDATE_EVENT]) {
      LOG(@"sendDeviceEventWithName: %@", eventName);
    }
    [self sendEventWithName:eventName body:body];
  } else {
    [self queueEventWithName:eventName body:body];
  }
}

- (void)setIsReactReady:(BOOL)isReactReady {
  if (_isReactReady) {
    LOG(@"received ReactReady notification after ready");
    return;
  }
  _isReactReady = isReactReady;
  // PurgeEvents must happen after isReactReady, however, I'm not positive this is truly thread-safe.
  // If a notification is queued during PurgeEvents, there could be an execption
  [self purgeEvents];
}

- (void)queueEventWithName:(NSString *)eventName body:(id)body {
  LOG(@"Queued Event: %@", eventName);
  OOQueuedEvent *event = [[OOQueuedEvent alloc] initWithWithName:eventName body:body];
  [self.queuedEvents addObject:event];
}

- (void)purgeEvents {
  LOG(@"Purging Events to skin");
  for (OOQueuedEvent *event in self.queuedEvents) {
    [self sendDeviceEventWithName:event.eventName body:event.body];
  }
  [self.queuedEvents removeAllObjects];
}

@end
