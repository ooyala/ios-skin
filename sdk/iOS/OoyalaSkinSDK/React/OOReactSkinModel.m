//
//  OOReactSkinModel.m
//  OoyalaSkinSDK
//
//  Created by Maksim Kupetskii on 8/13/18.
//  Copyright Â© 2018 ooyala. All rights reserved.
//

#import <React/RCTRootView.h>

#import <MediaPlayer/MPVolumeView.h>

#import "OOReactSkinModel.h"
#import "OOReactSkinBridge.h"
#import "OOReactSkinBridgeModuleMain.h"

#import "OOSkinViewControllerDelegate.h"
#import "OOSkinOptions.h"
#import "OOSkinPlayerObserver.h"
#import "OOUpNextManager.h"
#import "NSDictionary+Utils.h"
#import "UIColor+Utils.h"

#import <OoyalaSDK/OOOoyalaPlayer.h>
#import <OoyalaSDK/OODebugMode.h>
#import <OoyalaSDK/OODiscoveryManager.h>
#import <OoyalaSDK/OOAudioTrackProtocol.h>
#import <OoyalaSDK/OOAudioSession.h>
#import <OoyalaSDK/OOClosedCaptionsStyle.h>
#import <OoyalaSDK/OOVideo.h>

@interface OOReactSkinModel () <OOReactSkinBridgeDelegate, OOReactSkinModelDelegate, OOAudioSessionDelegate>

@property (nonatomic, weak) id<OOSkinViewControllerDelegate> skinControllerDelegate;

@property (nonatomic, weak) OOOoyalaPlayer *player;
@property (nonatomic, weak) OOSkinOptions *skinOptions;
@property (nonatomic) OOSkinPlayerObserver *playerObserver;
@property (nonatomic) OOUpNextManager *upNextManager;
@property (nonatomic, readwrite) RCTBridge *bridge;

@end

@implementation OOReactSkinModel

#pragma mark - Constants
#pragma mark Events
static NSString *discoveryResultsReceived = @"discoveryResultsReceived";
static NSString *ccStylingChanged         = @"ccStylingChanged";

#pragma mark Keys
static NSString *upNextKey              = @"upNext";
static NSString *volumeKey              = @"volume";
static NSString *textSizeKey            = @"textSize";
static NSString *textColorKey           = @"textColor";
static NSString *backgroundColorKey     = @"backgroundColor";
static NSString *textBackgroundColorKey = @"textBackgroundColor";
static NSString *backgroundOpacityKey   = @"backgroundOpacity";
static NSString *textOpacityKey         = @"textOpacity";
static NSString *fontNameKey            = @"fontName";
static NSString *audioKey               = @"audio";
static NSString *audioLanguageKey       = @"audioLanguage";
static NSString *embed_CodeKey          = @"embed_code";
static NSString *embedCodeKey           = @"embedCode";
static NSString *nameKey                = @"name";
static NSString *previewImgUrlKey       = @"preview_image_url";
static NSString *durationKey            = @"duration";
static NSString *bucket_InfoKey         = @"bucket_info";
static NSString *bucketInfoKey          = @"bucketInfo";
static NSString *descriptionKey         = @"description";
static NSString *imageUrlKey            = @"imageUrl";
static NSString *resultsKey             = @"results";
static NSString *volumePropertyKey      = @"outputVolume";
static NSString *volumeChangeKey        = @"volumeChanged";

#pragma mark - Init
- (instancetype)initWithWithPlayer:(OOOoyalaPlayer *)player
                       skinOptions:(OOSkinOptions *)skinOptions
            skinControllerDelegate:(id<OOSkinViewControllerDelegate>)skinControllerDelegate {
  if (self = [super init]) {
    _player = player;
    _skinOptions = skinOptions;
    _skinControllerDelegate = skinControllerDelegate;
    _skinConfig = [NSDictionary dictionaryFromSkinConfigFile:_skinOptions.configFileName
                                                  mergedWith:_skinOptions.overrideConfigs];
    _bridge = [[OOReactSkinBridge alloc] initWithDelegate:self launchOptions:nil];

    OOAudioSession.sharedInstance.delegate = self;

    _playerObserver = [[OOSkinPlayerObserver alloc] initWithPlayer:player ooReactSkinModel:self];
    _upNextManager = [[OOUpNextManager alloc] initWithPlayer:self.player
                                            ooReactSkinModel:self
                                                      config:self.skinConfig[upNextKey]];

    // Audio settings
    [self setupAudioSettingsFromConfig:_skinConfig];
  }
  return self;
}

- (void)dealloc {
}

#pragma mark - Public

- (RCTRootView *)viewForModuleWithName:(NSString *)moduleName{
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:self.bridge
                                                   moduleName:moduleName
                                            initialProperties:self.skinConfig];
  return rootView;
}

- (void)sendEventWithName:(NSString *)eventName body:(id)body {
  [self.bridge.skinEventsEmitter sendDeviceEventWithName:eventName body:body];
}

- (void)setIsReactReady:(BOOL)isReactReady {
  self.bridge.skinEventsEmitter.isReactReady = isReactReady;
  [self sendEventWithName:volumeChangeKey
                     body:@{volumeKey: @(OOAudioSession.sharedInstance.applicationVolume)}];
}

- (void)ccStyleChanged:(NSNotification *)notification {
  self.closedCaptionsDeviceStyle = [OOClosedCaptionsStyle new];
  NSNumber *textSize                    = @(self.closedCaptionsDeviceStyle.textSize);
  UIColor *textColor                    = self.closedCaptionsDeviceStyle.textColor;
  UIColor *backgroundColor              = self.closedCaptionsDeviceStyle.windowColor;
  UIColor *textBackgroundColor          = self.closedCaptionsDeviceStyle.backgroundColor;
  NSString *fontName                    = self.closedCaptionsDeviceStyle.textFontName;
  NSNumber *textOpacity                 = @(self.closedCaptionsDeviceStyle.textOpacity);
  NSNumber *backgroundOpacity           = @(self.closedCaptionsDeviceStyle.backgroundOpacity);
  NSString *backgroundColorHexValue     = [UIColor hexStringFromColor:backgroundColor];
  NSString *textBackgroundColorHexValue = [UIColor hexStringFromColor:textBackgroundColor];
  NSString *textColorHexValue           = [UIColor hexStringFromColor:textColor];
//  MACaptionAppearanceTextEdgeStyle edgeStyle = self.closedCaptionsDeviceStyle.edgeStyle;

  NSDictionary *params = @{textSizeKey:            textSize,
                           textColorKey:           textColorHexValue,
                           backgroundColorKey:     backgroundColorHexValue,
                           textBackgroundColorKey: textBackgroundColorHexValue,
                           backgroundOpacityKey:   backgroundOpacity,
                           textOpacityKey:         textOpacity,
                           fontNameKey:            fontName};
  [self sendEventWithName:ccStylingChanged body:params];
}

- (CGRect)videoViewFrame {
  return self.skinControllerDelegate.videoViewFrame;
}

- (void)setReactViewInteractionEnabled:(BOOL)reactViewInteractionEnabled {
  self.skinControllerDelegate.reactViewInteractionEnabled = reactViewInteractionEnabled;
}

#pragma mark - Private

- (void)setupAudioSettingsFromConfig:(NSDictionary *)config {
  NSDictionary *audioSettingsJSON    = config[audioKey];
  NSString *defaultAudioLanguageCode = audioSettingsJSON[audioLanguageKey];

  if (defaultAudioLanguageCode) {
    [self.player setDefaultConfigAudioTrackLanguageCode:defaultAudioLanguageCode];
  }
}

#pragma mark - Discovery UI

- (void)maybeLoadDiscovery:(NSString *)embedCode {
  if (self.player.currentItem.embedCode && self.skinOptions.discoveryOptions) {
    [OODiscoveryManager getResults:self.skinOptions.discoveryOptions
                         embedCode:embedCode
                             pcode:self.player.pcode
                        parameters:nil
                          callback:^(NSArray *results, OOOoyalaError *error) {
      if (error) {
        LOG(@"discovery request failed, error is %@", error.description);
      } else {
        [self handleDiscoveryResults:results embedCode:embedCode];
      }
    }];
  }
}

- (void)handleDiscoveryResults:(NSArray *)results embedCode:(NSString *)currentEmbedCode {
  NSMutableArray *discoveryArray = [NSMutableArray new];
  for (NSDictionary *dict in results) {
    NSString *embedCode = dict[embed_CodeKey];
    if ([embedCode isEqualToString:currentEmbedCode]) { continue; }
    NSString *name              = dict[nameKey];
    NSString *imageUrl          = dict[previewImgUrlKey];
    NSNumber *duration          = @([dict[durationKey] doubleValue] / 1000);
    NSString *bucketInfo        = dict[bucket_InfoKey];
    NSString *description       = dict[descriptionKey];// we assume we always get a string description, even if it is empty ("")
    NSDictionary *discoveryItem = @{nameKey:        name,
                                    embedCodeKey:   embedCode,
                                    imageUrlKey:    (NSNull *)imageUrl != NSNull.null ? imageUrl : @"",
                                    durationKey:    duration,
                                    bucketInfoKey:  bucketInfo,
                                    descriptionKey: description};
    [discoveryArray addObject:discoveryItem];
  }
  if (discoveryArray.count > 0 && discoveryArray[0]) {
    [self.upNextManager setNextVideo:discoveryArray[0]];
  }
  NSDictionary *eventBody = @{resultsKey: discoveryArray};
  [self sendEventWithName:discoveryResultsReceived body:eventBody];
}


#pragma mark - RCTBridgeDelegate

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge {
  return self.skinOptions.jsCodeLocation;
}

#pragma mark - OOReactSkinBridgeDelegate
// Binding ooReactSkinModel as a delegate for handling RN events
- (void)bridge:(OOReactSkinBridge *)bridge didLoadModule:(id<OOReactSkinBridgeModule>)module {
  if ([module isKindOfClass:OOReactSkinBridgeModuleMain.class]) {
    [(OOReactSkinBridgeModuleMain *)module setSkinViewDelegate:self];
  }
}

#pragma mark - OOReactSkinModelDelegate

- (void)handleAudioTrackSelection:(NSString *)audioTrackName {
  // TODO: Can we move this logic to core?
  for (id<OOAudioTrackProtocol> audioTrack in self.player.availableAudioTracks) {
    // This check is a temporary solution
    // Need to wait conclusion about tracks names (how need create them) from Ooyala
    if ([audioTrack.title isEqualToString:audioTrackName]) {
      self.player.defaultAudioTrack = audioTrack;
      [self.player setAudioTrack:audioTrack];
      return;
    }
  }

  LOG(@"handleAudioTrackSelection - Can't find audio track");
}

- (void)handlePlaybackSpeedRateSelection:(nullable NSNumber *)selectedPlaybackSpeedRate {
  if (selectedPlaybackSpeedRate) {
    [self.player changePlaybackSpeedRate:selectedPlaybackSpeedRate.floatValue];
  } else {
    LOG(@"handlePlaybackSpeedRateSelection - selectedPlaybackSpeedRate invalid type");
  }
}

- (void)handleDiscoveryClick:(NSString *)bucketInfo embedCode:(NSString *)embedCode {
  [OODiscoveryManager sendClick:self.skinOptions.discoveryOptions
                     bucketInfo:bucketInfo
                          pcode:self.player.pcode
                     parameters:nil];
  [self.player setEmbedCode:embedCode];
  [self.player play];
}

- (void)handleDiscoveryImpress:(NSString *)bucketInfo {
  [OODiscoveryManager sendImpression:self.skinOptions.discoveryOptions
                          bucketInfo:bucketInfo
                               pcode:self.player.pcode
                          parameters:nil];
}

- (void)handleIconClick:(NSInteger)index {
  [self.player onAdIconClicked:index];
}

- (void)handleLanguageSelection:(NSString *)language {
  [self.player setClosedCaptionsLanguage:language];
}

- (void)handleLearnMore {
  [self.player clickAd];
}

- (void)handleMoreOption {
  [self.player pause];
}

- (void)handleOverlay:(NSString *)url {
  [self.player onAdOverlayClicked:url];
}

- (void)handlePip {
  [self.player togglePictureInPictureMode];
}

- (void)handlePlay {
  [self.player play];
}

- (void)handlePlayPause {
  self.player.state == OOOoyalaPlayerStatePlaying ? [self.player pause] : [self.player play];
}

- (void)playPauseFromAdTappedNotification {
  if (!self.skinControllerDelegate.isReactViewInteractionEnabled) {
    [self handlePlayPause];
  }
}

- (void)handleReplay {
  [self.player play];
}

- (void)handleRewind {
  dispatch_async(dispatch_get_main_queue(), ^{
    if (self.player) {
      Float64 playheadTime = self.player.playheadTime;
      Float64 seekBackTo   = playheadTime - 10;
      if (self.player.seekableTimeRange.start.value > seekBackTo ) {
        seekBackTo = 0;
      }
      [self.player seek:seekBackTo];
    }
  });
}

- (void)handleScrub:(Float64)position {
  if (self.player) {
    CMTimeRange seekableRange = self.player.seekableTimeRange;
    Float64 start             = CMTimeGetSeconds(seekableRange.start);
    Float64 duration          = CMTimeGetSeconds(seekableRange.duration);
    Float64 playhead          = position * duration + start;
    if (playhead < 0.0f) {
      playhead = 0.0f;
    }
    
    [self.player seek:playhead];
  }
}

- (void)handleSkip {
  [self.player skipAd];
}

- (void)handleSocialShare {
  [self.player pause];
}

- (void)handleTouch:(NSDictionary *)result {
  [NSNotificationCenter.defaultCenter postNotificationName:OOOoyalaPlayerHandleTouchNotification object:result];
}

- (void)handleUpNextClick {
  [self.upNextManager goToNextVideo];
}

- (void)handleUpNextDismiss {
  [self.upNextManager onDismissPressed];
}

- (void)handleVolumeChanged:(float)volume {
  MPVolumeView *volumeView = [MPVolumeView new];
  UISlider *volumeViewSlider;
  
  for (UIView *view in volumeView.subviews) {
    if ([view isKindOfClass:UISlider.class]) {
      volumeViewSlider = (UISlider *)view;
      break;
    }
  }
  
  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.01 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
    volumeViewSlider.value = volume;
  });
}

- (void)setEmbedCode:(NSString *)embedCode {
  [self.player setEmbedCode:embedCode];
  [self.player play];
}

- (void)toggleFullscreen {
  [self.skinControllerDelegate toggleFullscreen];
}

- (void)toggleStereoMode {
  [self.skinControllerDelegate toggleStereoMode];
}

#pragma mark - OOAudioSessionDelegate

- (void)volumeChanged:(float)volume {
  [self sendEventWithName:volumeChangeKey
                     body:@{volumeKey: @(volume)}];
}

@end
