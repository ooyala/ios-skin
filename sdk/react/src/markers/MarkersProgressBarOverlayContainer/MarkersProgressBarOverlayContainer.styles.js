// @flow

import { StyleSheet } from 'react-native';

export default StyleSheet.create({
  root: {
    height: '100%',
    left: 0,
    position: 'absolute',
    top: 0,
    width: '100%',
  },
});
