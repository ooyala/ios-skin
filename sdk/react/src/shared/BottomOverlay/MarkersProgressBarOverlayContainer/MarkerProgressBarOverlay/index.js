// @flow

import MarkerProgressBarOverlay from './MarkerProgressBarOverlay';

export default MarkerProgressBarOverlay;
