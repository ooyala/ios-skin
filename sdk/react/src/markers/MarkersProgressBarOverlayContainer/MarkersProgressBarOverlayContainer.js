// @flow

import * as React from 'react';
import { View } from 'react-native';
import type { ViewStyleProp } from 'react-native/Libraries/StyleSheet/StyleSheet';

import MarkerProgressBarOverlay from '../MarkerProgressBarOverlay';
import styles from './MarkersProgressBarOverlayContainer.styles';
import type { Marker as MarkerType } from '../../types/Marker';

type Props = {
  accentColor?: ?string,
  duration: number,
  markers: Array<MarkerType>,
  style?: ViewStyleProp,
};

const MarkersProgressBarOverlayContainer = ({
  accentColor, duration, markers, style,
}: Props) => (
  markers.length > 0
    ? (
      <View style={[styles.root, style]}>
        {markers.map((marker, index) => (
          <MarkerProgressBarOverlay
            accentColor={accentColor}
            duration={duration}
            key={index} // eslint-disable-line react/no-array-index-key
            marker={marker}
          />
        ))}
      </View>
    )
    : null
);

MarkersProgressBarOverlayContainer.defaultProps = {
  accentColor: undefined,
  style: undefined,
};

export default MarkersProgressBarOverlayContainer;
