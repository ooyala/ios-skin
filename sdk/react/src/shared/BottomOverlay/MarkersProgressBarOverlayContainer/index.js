// @flow

import MarkersProgressBarOverlayContainer from './MarkersProgressBarOverlayContainer';

export default MarkersProgressBarOverlayContainer;
