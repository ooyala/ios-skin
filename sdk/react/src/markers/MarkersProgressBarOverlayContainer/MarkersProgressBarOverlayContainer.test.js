// @flow

import React from 'react';
import TestRenderer from 'react-test-renderer';

import MarkersProgressBarOverlayContainer from './MarkersProgressBarOverlayContainer';

describe('MarkersProgressBarOverlayContainer', () => {
  it('renders matching snapshot', () => {
    const testRenderer = TestRenderer.create(
      <MarkersProgressBarOverlayContainer
        duration={60}
        markers={[
          {
            start: 'start',
            type: 'text',
            text: 'Hello, world!',
          },
          {
            iconUrl: 'http://example.com/icon.png',
            start: 'start',
            type: 'icon',
          },
        ]}
        onSeek={() => undefined}
      />,
    );

    expect(testRenderer.toJSON()).toMatchSnapshot();
  });

  it('does not render anything when markers array is empty', () => {
    const testRenderer = TestRenderer.create(
      <MarkersProgressBarOverlayContainer
        duration={60}
        markers={[]}
        onSeek={() => undefined}
      />,
    );

    expect(testRenderer.toJSON()).toBeNull();
  });
});
