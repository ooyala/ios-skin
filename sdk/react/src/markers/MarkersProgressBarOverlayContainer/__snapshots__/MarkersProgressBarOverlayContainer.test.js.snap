// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkersProgressBarOverlayContainer renders matching snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "bottom": 0,
          "minWidth": 10,
          "position": "absolute",
          "top": 0,
        },
        Object {
          "left": 0,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "bottom": 0,
          "minWidth": 10,
          "position": "absolute",
          "top": 0,
        },
        Object {
          "left": 0,
        },
      ]
    }
  />
</View>
`;
